// graphql-rate-limit/dist/esm/index.js
// src/lib/field-directive.ts
import {
  defaultFieldResolver,
  DirectiveLocation,
  GraphQLBoolean,
  GraphQLDirective,
  GraphQLInt,
  GraphQLList,
  GraphQLString
} from "graphql";
import {SchemaDirectiveVisitor} from "@graphql-tools/utils";

// src/lib/rate-limit-error.ts
var RateLimitError = class extends Error {
  constructor(message) {
    super(message);
    this.isRateLimitError = true;
    Object.setPrototypeOf(this, RateLimitError.prototype);
  }
};

// src/lib/get-graphql-rate-limiter.ts
import get from "lodash.get";
import ms from "ms";

// src/lib/in-memory-store.ts
var InMemoryStore = class {
  constructor() {
    this.state = {};
  }
  setForIdentity(identity, timestamps) {
    this.state = {
      ...this.state || {},
      [identity.contextIdentity]: {
        ...this.state[identity.contextIdentity] || {},
        [identity.fieldIdentity]: [...timestamps]
      }
    };
  }
  getForIdentity(identity) {
    const ctxState = this.state[identity.contextIdentity];
    return ctxState && ctxState[identity.fieldIdentity] || [];
  }
};

// src/lib/batch-request-cache.ts
var getNoOpCache = () => ({
  set: ({newTimestamps}) => newTimestamps
});
var getWeakMapCache = () => {
  const cache = new WeakMap();
  return {
    set: ({
      context,
      fieldIdentity,
      newTimestamps
    }) => {
      const currentCalls = cache.get(context) || {};
      currentCalls[fieldIdentity] = [
        ...currentCalls[fieldIdentity] || [],
        ...newTimestamps
      ];
      cache.set(context, currentCalls);
      return currentCalls[fieldIdentity];
    }
  };
};

// src/lib/get-graphql-rate-limiter.ts
var DEFAULT_WINDOW = 60 * 1e3;
var DEFAULT_MAX = 5;
var DEFAULT_FIELD_IDENTITY_ARGS = [];
var getFieldIdentity = (fieldName, identityArgs, resolveArgs) => {
  const argsKey = identityArgs.map((arg) => get(resolveArgs, arg));
  return [fieldName, ...argsKey].join(":");
};
var getGraphQLRateLimiter = (userConfig) => {
  const defaultConfig = {
    enableBatchRequestCache: false,
    formatError: ({fieldName}) => {
      return `You are trying to access '${fieldName}' too often`;
    },
    identifyContext: () => {
      throw new Error("You must implement a createRateLimitDirective.config.identifyContext");
    },
    store: new InMemoryStore()
  };
  const {enableBatchRequestCache, identifyContext, formatError, store} = {
    ...defaultConfig,
    ...userConfig
  };
  const batchRequestCache = enableBatchRequestCache ? getWeakMapCache() : getNoOpCache();
  const rateLimiter = async ({
    args,
    context,
    info
  }, {
    arrayLengthField,
    identityArgs,
    max,
    window,
    message,
    readOnly,
    uncountRejected
  }) => {
    const contextIdentity = identifyContext(context);
    const windowMs = window ? ms(window) : DEFAULT_WINDOW;
    const fieldIdentity = getFieldIdentity(info.fieldName, identityArgs || DEFAULT_FIELD_IDENTITY_ARGS, args);
    const maxCalls = max || DEFAULT_MAX;
    const callCount = arrayLengthField && get(args, [arrayLengthField, "length"]) || 1;
    const identity = {contextIdentity, fieldIdentity};
    const timestamp = Date.now();
    const newTimestamps = [...new Array(callCount || 1)].map(() => timestamp);
    const batchedTimestamps = batchRequestCache.set({
      context,
      fieldIdentity,
      newTimestamps
    });
    const filteredAccessTimestamps = (await store.getForIdentity(identity)).filter((t) => {
      return t + windowMs > Date.now();
    });
    const limitReached = filteredAccessTimestamps.length + batchedTimestamps.length > maxCalls;
    const timestampsToStore = [
      ...filteredAccessTimestamps,
      ...!uncountRejected || !limitReached ? batchedTimestamps : []
    ];
    if (!readOnly) {
      await store.setForIdentity(identity, timestampsToStore, windowMs);
    }
    const errorMessage = message || formatError({
      contextIdentity,
      fieldIdentity,
      fieldName: info.fieldName,
      max: maxCalls,
      window: windowMs
    });
    return limitReached ? errorMessage : void 0;
  };
  return rateLimiter;
};

// src/lib/field-directive.ts
var createRateLimitDirective = (customConfig = {}) => {
  const rateLimiter = getGraphQLRateLimiter(customConfig);
  class GraphQLRateLimit extends SchemaDirectiveVisitor {
    static getDirectiveDeclaration(directiveName) {
      return new GraphQLDirective({
        args: {
          arrayLengthField: {
            type: GraphQLString
          },
          identityArgs: {
            type: new GraphQLList(GraphQLString)
          },
          max: {
            type: GraphQLInt
          },
          message: {
            type: GraphQLString
          },
          window: {
            type: GraphQLString
          },
          uncountRejected: {
            type: GraphQLBoolean
          }
        },
        locations: [DirectiveLocation.FIELD_DEFINITION],
        name: directiveName
      });
    }
    visitFieldDefinition(field) {
      const {resolve = defaultFieldResolver} = field;
      field.resolve = async (parent, args, context, info) => {
        const errorMessage = await rateLimiter({
          parent,
          args,
          context,
          info
        }, this.args);
        if (errorMessage) {
          throw customConfig.createError ? customConfig.createError(errorMessage) : new RateLimitError(errorMessage);
        }
        return resolve(parent, args, context, info);
      };
    }
  }
  return GraphQLRateLimit;
};

// src/lib/store.ts
var Store = class {
};

// src/lib/redis-store.ts
var RedisStore = class {
  constructor(redisStoreInstance) {
    this.nameSpacedKeyPrefix = "redis-store-id::";
    this.generateNamedSpacedKey = (identity) => {
      return `${this.nameSpacedKeyPrefix}${identity.contextIdentity}:${identity.fieldIdentity}`;
    };
    this.store = redisStoreInstance;
  }
  setForIdentity(identity, timestamps, windowMs) {
    return new Promise((res, rej) => {
      const expiry = windowMs ? [
        "EX",
        Math.ceil((Date.now() + windowMs - Math.max(...timestamps)) / 1e3)
      ] : [];
      this.store.set([
        this.generateNamedSpacedKey(identity),
        JSON.stringify([...timestamps]),
        ...expiry
      ], (err) => {
        if (err)
          return rej(err);
        return res();
      });
    });
  }
  async getForIdentity(identity) {
    return new Promise((res, rej) => {
      this.store.get(this.generateNamedSpacedKey(identity), (err, obj) => {
        if (err) {
          return rej(err);
        }
        return res(obj ? JSON.parse(obj) : []);
      });
    });
  }
};

// src/lib/rate-limit-shield-rule.ts
import {rule} from "graphql-shield";
var createRateLimitRule = (config) => {
  const rateLimiter = getGraphQLRateLimiter(config);
  return (fieldConfig) => rule({cache: "no_cache"})(async (parent, args, context, info) => {
    const errorMessage = await rateLimiter({
      parent,
      args,
      context,
      info
    }, fieldConfig);
    if (errorMessage) {
      return config.createError ? config.createError(errorMessage) : new RateLimitError(errorMessage);
    }
    return true;
  });
};

// src/index.ts
var src_default = createRateLimitDirective;
export {
  InMemoryStore,
  RateLimitError,
  RedisStore,
  Store,
  createRateLimitDirective,
  createRateLimitRule,
  src_default as default,
  getFieldIdentity,
  getGraphQLRateLimiter
};
